File: Part 8/NTSA_timefreq_filterHilbert.m
%%
%     COURSE: Solved problems in neural time series analysis
%    SECTION: Time-frequency analyses
%      VIDEO: Narrowband filtering and the Hilbert transform
% Instructor: sincxpress.com
%
%%


%% design FIR filter via fir1

srate   = 1000;
nyquist = srate/2;
frange  = [10 15];
order   = round( 20*srate/frange(1) );

% filter kernel
filtkern = fir1(order,frange/nyquist);

% compute the power spectrum of the filter kernel
filtpow = abs(fft(filtkern)).^2;
% compute the frequencies vector and remove negative frequencies
hz      = linspace(0,srate/2,floor(length(filtkern)/2)+1);
filtpow = filtpow(1:length(hz));



figure(1), clf
subplot(131)
plot(filtkern,'linew',2)
xlabel('Time points')
title('Filter kernel (fir1)')
axis square


% plot amplitude spectrum of the filter kernel
subplot(132), hold on
plot(hz,filtpow,'ks-','linew',2,'markerfacecolor','w')

plot([0 frange(1) frange frange(2) nyquist],[0 0 1 1 0 0],'ro-','linew',2,'markerfacecolor','w')

% dotted line corresponding to the lower edge of the filter cut-off
plot([1 1]*frange(1),get(gca,'ylim'),'k:')

% make the plot look nicer
set(gca,'xlim',[0 frange(1)*4])%,'ylim',[-.05 1.05])
xlabel('Frequency (Hz)'), ylabel('Filter gain')
legend({'Actual';'Ideal'})
title('Frequency response of filter (fir1)')
axis square


subplot(133), hold on
plot(hz,10*log10(filtpow),'ks-','linew',2,'markersize',10,'markerfacecolor','w')
plot([1 1]*frange(1),get(gca,'ylim'),'k:')
set(gca,'xlim',[0 frange(1)*4],'ylim',[-80 2])
xlabel('Frequency (Hz)'), ylabel('Filter gain (dB)')
title('Frequency response of filter (fir1)')
axis square

%% apply the filter to the data

pnts = 10000;
origdata = randn(pnts,1);
timevec = (0:pnts-1)/srate;

filtdat = filtfilt(filtkern,1,origdata);

%% Hilbert transform

hildat = hilbert(filtdat);

figure(2), clf
subplot(311), hold on
plot(timevec,filtdat,'k')
plot(timevec,real(hildat),'ro')
zoom on
title('Filtered signal')
legend({'Original';'Real'})

subplot(312)
plot(timevec,real(hildat), timevec,imag(hildat))
title('Complex representation')
legend({'Real';'Imag'})

subplot(313)
plot(timevec,abs(hildat))
xlabel('Time (s)'), ylabel('Amplitude')
title('Band-limited amplitude envelope')

%% done.

File: Part 8/NTSA_timefreq_filterHilbert.py
# %%
#     COURSE: Solved problems in neural time series analysis
#    SECTION: Time-frequency analyses
#    TEACHER: Mike X Cohen, sincxpress.com
#      VIDEO: Narrowband filtering and the Hilbert transform

# !%matplotlib inline
import os
import sys

import matplotlib.pyplot as plt
import numpy as np
from numpy.fft import fft, ifft
from scipy.signal import filtfilt, firwin, freqz, hilbert

# Set figure settings
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from utils import get_cmap, set_fig_dpi, set_style

set_fig_dpi(), set_style()
cmap = get_cmap('parula')

# %% Design FIR filter via firwin
srate = 1000
nyquist = srate / 2
freq_range = [10, 15]
order = round(20 * srate / freq_range[0])
order |= 1  # Ensure that order is odd

# Filter kernel
filt_kern = firwin(order + 1, freq_range, nyq=nyquist, pass_zero=False)

# Compute the power spectrum of the filter kernel
filt_pow = np.abs(np.fft.fft(filt_kern)) ** 2
# Compute the frequencies vector and remove negative frequencies
hz = np.linspace(0, srate / 2, int(np.floor(len(filt_kern) / 2)) + 1)
filt_pow = filt_pow[: len(hz)]

# %% Plot filter characteristics

plt.figure(figsize=(15, 5))
plt.subplot(131)
plt.plot(filt_kern)
plt.xlabel('Time Points')
plt.title('Filter Kernel (firwin)')

plt.subplot(132)
plt.plot(hz, filt_pow, 'ks-', markerfacecolor='w', markersize=8)
plt.plot(
    [0, freq_range[0], freq_range[0], freq_range[1], freq_range[1], nyquist],
    [0, 0, 1, 1, 0, 0],
    'ro-',
    markerfacecolor='w',
    label='Actual',
)
plt.plot([freq_range[0], freq_range[0]], plt.gca().get_ylim(), 'k:', label='Ideal')
plt.xlim(0, freq_range[0] * 4)
plt.xlabel('Frequency (Hz)')
plt.ylabel('Filter gain')
plt.legend()
plt.title('Frequency Response of Filter (firwin)')

plt.subplot(133)
plt.plot(hz, 10 * np.log10(filt_pow), 'ks-', markersize=10, markerfacecolor='w')
plt.plot([freq_range[0], freq_range[0]], plt.gca().get_ylim(), 'k:')
plt.xlim(0, freq_range[0] * 4)
plt.ylim(-80, 2)
plt.xlabel('Frequency (Hz)')
plt.ylabel('Filter Gain (dB)')
plt.title('Frequency Response of Filter (dB)')

plt.tight_layout()
plt.show()

# %% Apply the filter to the data
pnts = 10000
orig_data = np.random.randn(pnts)
time_vec = np.arange(pnts) / srate

filt_data = filtfilt(filt_kern, 1, orig_data)

# %% Hilbert transform

hil_data = hilbert(filt_data)

plt.figure(figsize=(10, 8))
plt.subplot(311)
plt.plot(time_vec, filt_data, 'k', label='Original')
plt.plot(time_vec, np.real(hil_data), 'ro', label='Real')
plt.title('Filtered signal')
plt.legend()

plt.subplot(312)
plt.plot(time_vec, np.real(hil_data), label='Real')
plt.plot(time_vec, np.imag(hil_data), label='Imag')
plt.plot(time_vec, np.abs(hil_data), 'k', label='Envelope')
plt.title('Complex Representation')
plt.legend()

plt.subplot(313)
plt.plot(time_vec, np.abs(hil_data), 'k')
plt.xlabel('Time (s)')
plt.ylabel('Amplitude')
plt.title('Band-Limited Amplitude Envelope')

plt.tight_layout()
plt.show()

# %% Done.
